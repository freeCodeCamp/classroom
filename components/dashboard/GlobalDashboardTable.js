import React from 'react';
import { useTable } from 'react-table';
import getStudentActivityAndProgress from '../../util/processDashboardData';
export default function GlobalDashboardTable({
  classroomId,
  studentData,
  certifications
}) {
  let formattedData = React.useMemo(() => [], []);
  // formattedData = [];
  studentData.forEach(data => {
    // The 0th index is the all the certifications that the student is enrolled in
    // The 1st index is all available certifications by FCC
    let studentCertsAndAllCerts = [data.certifications, certifications];
    let activityAndProgressResults = getStudentActivityAndProgress(
      studentCertsAndAllCerts
    );
    let summary = {
      name: data.email,
      activity: activityAndProgressResults[0],
      progress: activityAndProgressResults[1],
      detail: (
        <a
          // TODO:
          href={`/dashboard/v2/details/${classroomId}/` + `${data.name}`}
        >
          {' '}
          details{' '}
        </a>
      )
    };
    formattedData.push(summary);
  });

  const mapData = function (original_data) {
    let table_data = original_data.map(student => {
      let mapped_student = {
        col1: student.name,
        col2: student.activity,
        col3: student.progress,
        col4: student.detail
      };
      return mapped_student;
    });
    return table_data;
  };

  const data = React.useMemo(() => mapData(formattedData), [formattedData]);
  const columns = React.useMemo(
    () => [
      {
        Header: 'Student Name',
        accessor: 'col1', // accessor is the "key" in the data
        width: '20%'
      },
      {
        Header: 'Activity',
        accessor: 'col2',
        width: '10%'
      },
      {
        Header: 'Progress',
        accessor: 'col3',
        width: '20%'
      },
      {
        Header: 'Details',
        accessor: 'col4',
        width: '50%'
      }
    ],
    []
  );

  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    useTable({ columns, data });

  return (
    <>
      <table
        {...getTableProps()}
        style={{ border: 'solid 1px #0a0a23', width: '100%', margin: 'auto' }}
      >
        <thead>
          {headerGroups.map((headerGroup, index) => (
            <tr {...headerGroup.getHeaderGroupProps()} key={index}>
              {headerGroup.headers.map((column, index) => (
                <th
                  {...column.getHeaderProps()}
                  style={{
                    borderBottom: 'solid 3px grey',
                    color: 'black',
                    fontWeight: 'bold'
                  }}
                  key={index}
                >
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map((row, index) => {
            prepareRow(row);
            return (
              <tr {...row.getRowProps()} key={index}>
                {row.cells.map((cell, index) => {
                  return (
                    <td
                      {...cell.getCellProps()}
                      style={{
                        padding: '10px',
                        border: 'solid 1px grey',
                        textAlign: 'center',
                        width: cell.column.width
                      }}
                      key={index}
                    >
                      {cell.render('Cell')}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
    </>
  );
}
