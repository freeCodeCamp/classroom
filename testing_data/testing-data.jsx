export const studentData = [
  {
    email: 'student[A]@gmail.com',
    certifications: [
      {
        '2022/responsive-web-design': {
          blocks: [
            {
              'learn-basic-css-by-building-a-cafe-menu': {
                completedChallenges: [
                  {
                    id: '5f33071498eb2472b87ddee4',
                    challengeName: 'Step 1',
                    completedDate: 1475094716730,
                    files: []
                  },
                  {
                    id: '5f3313e74582ad9d063e3a38',
                    challengeName: 'Step 2',
                    completedDate: 1537207306322,
                    files: []
                  }
                ]
              }
            }
          ]
        }
      },
      {
        'quality-assurance': {
          blocks: [
            {
              'advanced-node-and-express': {
                completedChallenges: [
                  {
                    id: '5895f700f9fc0f352b528e63',
                    challengeName: 'Set up a Template Engine',
                    completedDate: 98448684,
                    files: []
                  },
                  {
                    id: '5895f70df9fc0f352b528e6a',
                    challengeName: 'Create New Middleware',
                    completedDate: 98448643284,
                    files: []
                  }
                ]
              }
            },
            {
              'quality-assurance-and-testing-with-chai': {
                completedChallenges: [
                  {
                    id: '587d824a367417b2b2512c46',
                    challengeName: 'Learn How JavaScript Assertions Work',
                    completedDate: 47664591,
                    files: []
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  },
  {
    email: 'student[B]@gmail.com',
    certifications: [
      {
        'quality-assurance': {
          blocks: [
            {
              'advanced-node-and-express': {
                completedChallenges: [
                  {
                    id: '5895f700f9fc0f352b528e63',
                    challengeName: 'Set up a Template Engine',
                    completedDate: 98448684,
                    files: []
                  },
                  {
                    id: '5895f70df9fc0f352b528e6a',
                    challengeName: 'Create New Middleware',
                    completedDate: 98448643284,
                    files: []
                  }
                ]
              }
            },
            {
              'quality-assurance-and-testing-with-chai': {
                completedChallenges: [
                  {
                    id: '587d824a367417b2b2512c46',
                    challengeName: 'Learn How JavaScript Assertions Work',
                    completedDate: 47664591,
                    files: []
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
];
export const certifications = [
  [
    {
      name: 'Learn HTML By Building a Cat Photo App',
      selector: 'learn-html-by-building-a-cat-photo-app',
      dashedName: 'learn-html-by-building-a-cat-photo-app',
      allChallenges: [
        '5dc174fcf86c76b9248c6eb2',
        '5dc1798ff86c76b9248c6eb3',
        // ... other challenge IDs
        '5ef9b03c81a63668521804ec',
        '5ef9b03c81a63668521804ee',
        '62bb4009e3458a128ff57d5d'
      ],
      order: 0
    },
    // ... other certifications
    {
      name: 'Build a Personal Portfolio Webpage Project',
      selector: 'build-a-personal-portfolio-webpage-project',
      dashedName: 'build-a-personal-portfolio-webpage-project',
      allChallenges: ['bd7158d8c242eddfaeb5bd13'],
      order: 19
    }
  ]
];

export const classroomId = 'cljbqrfoa0006kohk894ym6v3';
export const userId = 'cljyeiqeq0000e1k5fyxg20pa';


export const timestamps = [
  {
    name: 'student[A]@gmail.com',
    completedTimestamps: [
      1475094716730, 1537207306322, 98448684, 98448643284, 47664591
    ]
  },
  {
    name: 'student[B]@gmail.com',
    completedTimestamps: [98448684, 98448643284, 47664591]
  }
];

export const totalChallenges = 6;
