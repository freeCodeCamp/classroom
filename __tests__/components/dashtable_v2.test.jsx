import { render } from '@testing-library/react';
import React from 'react';
import GlobalDashboardTable from '../../components/dashtable_v2.js';

import {studentData, certifications, classroomId, timestamps, totalChallenges} from '../../testing_data/testing-data';

describe('GlobalDashboardTable', () => {
  // Define a fixed time in milliseconds
  const fixedTime = new Date('2016-03-09T12:00:00Z').getTime();

  // Spy on Date.prototype.getTime and mock its implementation
  const getTimeSpy = jest.spyOn(Date.prototype, 'getTime').mockImplementation(() => fixedTime);

  // Restore the original getTime method after the test
  afterEach(() => {
    getTimeSpy.mockRestore();
  });

  test('renders dashtable with correct data format', () => {
    const { container } = render(
      <GlobalDashboardTable
        studentData={studentData}
        certifications={certifications}
        classroomId={classroomId}
        totalChallenges={totalChallenges}
        timestamps={timestamps}
      />
    );
    expect(getTimeSpy).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
