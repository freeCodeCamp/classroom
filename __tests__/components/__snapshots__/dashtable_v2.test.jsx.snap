// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalDashboardTable renders dashtable with correct data format 1`] = `
<div>
  <table
    role="table"
    style="border: 1px solid #0a0a23; width: 100%; margin: auto;"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
          style="border-bottom: 3px solid grey; color: black; font-weight: bold;"
        >
          Student Email
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="border-bottom: 3px solid grey; color: black; font-weight: bold;"
        >
          Activity
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="border-bottom: 3px solid grey; color: black; font-weight: bold;"
        >
          Progress
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="border-bottom: 3px solid grey; color: black; font-weight: bold;"
        >
          Details
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 20%;"
        >
          student[A]@gmail.com
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 10%;"
        >
          <div
            class="bg-green-600 h-5 w-5"
            style="margin: auto;"
            title="Last completion time: 9/17/2018, 11:01:46 AM"
          />
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 20%;"
        >
          <div>
            <label>
              5
              /
              6
               
            </label>
            <meter
              id="progress"
              max="6"
              min="0"
              value="5"
            />
          </div>
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 50%;"
        >
          <a
            href="/dashboard/v2/details/cljbqrfoa0006kohk894ym6v3/student[A]@gmail.com"
          >
             
            details
             
          </a>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 20%;"
        >
          student[B]@gmail.com
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 10%;"
        >
          <div
            class="bg-red-600 h-5 w-5"
            style="margin: auto;"
            title="Last completion time: 2/13/1973, 2:50:43 AM"
          />
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 20%;"
        >
          <div>
            <label>
              3
              /
              6
               
            </label>
            <meter
              id="progress"
              max="6"
              min="0"
              value="3"
            />
          </div>
        </td>
        <td
          role="cell"
          style="padding: 10px; border: 1px solid grey; text-align: center; width: 50%;"
        >
          <a
            href="/dashboard/v2/details/cljbqrfoa0006kohk894ym6v3/student[B]@gmail.com"
          >
             
            details
             
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
